### PR TITLE
Upgraded to gulp 4 and moved React and Sass to separate files.

### PR DESCRIPTION
**Pull request by zinigor**
While trying to fix the Beta Builder I refactored the gulpfile thinking that the problem was there. It wasn't, but letting the code go to waste is going to be a shame, so here it is.

#### Changes proposed in this Pull Request:
* Upgrades gulp to version 4.
* Redefines tasks to use series and parallels instead of dependencies, makes for easier management.
* Moves React and SASS tasks to separate files, makes for easier readability.
* Adds a forgotten Search widget CSS file back into the glob.
* Fixes static.html file not being generated in watch mode.
* Fixes static.html file silently failing on errors.

#### Testing instructions:
* Test various `yarn` commands that we generally use for development and deployment.
* See if the admin and client area of Jetpack is fine.

### CODE REVIEW

**Comment by oskosk** [comment]
This one needs a rebase @zinigor !

**Comment by brbrr** [comment]
@zinigor Is this still something that you want to be merged?

**Comment by zinigor** [comment]
Yes it is, I'l get back to it! Thanks for the reminder!

**Comment by ockham** [comment]
@zinigor Are you planning to resuscitate this? I stumbled over [some unexpected behavior](https://stackoverflow.com/questions/37664138/gulp-globbing-excluding-files-then-unexcluding-not-working-as-described) of `gulp.src()` excludes which should be fixed in the `vinyl-fs version that gulp v4 uses...

**Comment by zinigor** [comment]
Oh, that gives this PR more urgency :) I'll try to work on it this week, thanks!

**Comment by ockham** [comment]
Missing a closing paren on L180:

```
SyntaxError: /home/bernie/src/jetpack/gulpfile.babel.js: Unexpected token, expected , (180:3)
```

**Comment by zinigor** [comment]
Yeah, sorry about that - I fixed it, but forgot to `git add` when amending a merge commit. Should be fixed now.

File: gulpfile.babel.js [hunk 1]
```diff
@@ -1,334 +1,43 @@
...
  * External dependencies
  */
-import autoprefixer from 'gulp-autoprefixer';
 import babel from 'gulp-babel';
-import banner from 'gulp-banner';
 import check from 'gulp-check';
-import cleanCSS from 'gulp-clean-css';
 import colors from 'ansi-colors';
-import del from 'del';
-import deleteLines from 'gulp-delete-lines';
```
Comment (ockham) [inline gulpfile.babel.js]: Seems like we're still using this in the `languages:build` task, no?

File: gulpfile.babel.js [hunk 2]
```diff
@@ -1,334 +1,43 @@
...
 import eslint from 'gulp-eslint';
 import i18n_calypso from 'i18n-calypso/cli';
 import jshint from 'gulp-jshint';
 import json_transform from 'gulp-json-transform';
+import log from 'fancy-log';
 import phplint from 'gulp-phplint';
 import phpunit from 'gulp-phpunit';
-import PluginError from 'plugin-error';
 import po2json from 'gulp-po2json';
-import rename from 'gulp-rename';
```
Comment (ockham) [inline gulpfile.babel.js]: Same as https://github.com/Automattic/jetpack/pull/9042/files#r225468922

**Comment by zinigor** [comment]
Thanks for taking a look, Bernie! I have pushed a fix for missing import statements.

File: gulpfile.babel.js [hunk 3]
```diff
@@ -1,334 +1,45 @@
 /**
  * External dependencies
  */
-import autoprefixer from 'gulp-autoprefixer';
 import babel from 'gulp-babel';
-import banner from 'gulp-banner';
 import check from 'gulp-check';
-import cleanCSS from 'gulp-clean-css';
 import colors from 'ansi-colors';
+import fs from 'fs';
```
Comment (ockham) [inline gulpfile.babel.js]: Nitpick -- out of alphabetical order :grin:

Comment (ockham) [reply gulpfile.babel.js]: Oh, so is `eslint` a bit further down. Nevermind I guess.

Comment (zinigor) [reply gulpfile.babel.js]: In for a penny - in for a pound :)

File: gulpfile.babel.js [hunk 4]
```diff
@@ -1,334 +1,45 @@
...
-				'Warning: gulp was unable to update static HTML files.\n\n' +
-				'If this is happening during watch, this warning is OK to dismiss: sometimes webpack fires watch handlers when source code is not yet built.'
-			) );
-		}
-	} );
-}
+import {} from './tools/builder/frontend-css';
+import {} from './tools/builder/admin-css';
+import {} from './tools/builder/react';
+import {} from './tools/builder/sass';
```
Comment (ockham) [inline gulpfile.babel.js]: :heart: the modularization!

Can we make the imports a bit less magical tho? :smile: What about

```difff
diff --git a/gulpfile.babel.js b/gulpfile.babel.js
index c87e3e6e3..31d79b54c 100644
--- a/gulpfile.babel.js
+++ b/gulpfile.babel.js
@@ -33,7 +33,7 @@ module.exports = {
 	alwaysIgnoredPaths: alwaysIgnoredPaths
 };
 
-import {} from './tools/builder/frontend-css';
+import frontendcss from './tools/builder/frontend-css';
 import {} from './tools/builder/admin-css';
 import {} from './tools/builder/react';
 import {} from './tools/builder/sass';
@@ -290,7 +290,7 @@ gulp.task( 'gutenpack:jetpack-blocks', function() {
 
 gulp.task(
 	'old-styles',
-	gulp.parallel( 'frontendcss', 'admincss', 'admincss:rtl', 'sass:old' )
+	gulp.parallel( frontendcss, 'admincss', 'admincss:rtl', 'sass:old' )
 );
 gulp.task( 'jshint', gulp.parallel( 'js:hint' ) );
 gulp.task( 'php', gulp.parallel( 'php:lint', 'php:unit' ) );
diff --git a/tools/builder/frontend-css.js b/tools/builder/frontend-css.js
index 55528191a..c39c0104f 100644
--- a/tools/builder/frontend-css.js
+++ b/tools/builder/frontend-css.js
@@ -75,7 +75,7 @@ const pathModifier = function( file, contents ) {
 };
 
 // Frontend CSS.  Auto-prefix and minimize.
-gulp.task( 'frontendcss', function() {
+export default function frontendcss() {
 	return gulp.src( concat_list )
 		.pipe( modify( { fileModifier: pathModifier } ) )
 		.pipe( autoprefixer(
@@ -101,9 +101,9 @@ gulp.task( 'frontendcss', function() {
 		.on( 'end', function() {
 			util.log( 'Front end modules CSS finished.' );
 		} );
-} );
+}
 
-gulp.task( 'frontendcss:separate', gulp.series( 'frontendcss', function() {
+gulp.task( 'frontendcss:separate', gulp.series( frontendcss, function() {
 	return gulp.src( separate_list )
 		.pipe( modify( { fileModifier: pathModifier } ) )
 		.pipe( autoprefixer(
```

See e.g. https://github.com/gulpjs/gulp/blob/master/docs/recipes/exports-as-tasks.md

File: tools/builder/frontend-css.js [hunk 5]
```diff
@@ -103,7 +103,7 @@ gulp.task( 'frontendcss', [ 'frontendcss:separate' ], function() {
 		} );
 } );
 
-gulp.task( 'frontendcss:separate', function() {
+gulp.task( 'frontendcss:separate', gulp.series( 'frontendcss', function() {
```
Comment (ockham) [inline tools/builder/frontend-css.js]: Oh BTW how come we're flipping dependencies here? (Previously, `frontendcss` depended on `frontendcss:separate`, now it's the other way round.)

File: gulpfile.babel.js (hunk 4 above)
Comment (ockham) [reply gulpfile.babel.js]: (Probably fine to do in a separate PR)

**Comment by zinigor** [comment]
Thanks for reviewing, Bernie! I have added explicit import statements that should be a good start for us. It proves to be a good pattern for separating concerns, but it's really a rabbit hole :)

**Comment by ockham** [comment]
I ran `yarn build-languages`, which eventually errored:

```
[...]
[19:36:56] ~/src/jetpack/languages/jetpack-zh_TW.mo
[19:36:56] ~/src/jetpack/languages/jetpack-zh_TW.po
[19:36:56] Finished '<anonymous>' after 1.01 s
[19:36:56] Finished 'languages:cleanup' after 6.85 min
[19:36:56] '<anonymous>' errored after 5.17 s
[19:36:56] Error: ENOENT: no such file or directory, open '~/src/jetpack/languages/jetpack-sv_SE.po'
[19:36:56] 'languages:build' errored after 6.85 min
[19:36:56] 'languages' errored after 6.85 min
error Command failed with exit code 1.
```

I'm guessing that maybe the `languages` task should be a `series` rather than `parallel`?

File: gulpfile.babel.js [hunk 6]
```diff
@@ -1,334 +1,46 @@
...
-				'If this is happening during watch, this warning is OK to dismiss: sometimes webpack fires watch handlers when source code is not yet built.'
-			) );
-		}
-	} );
-}
+import frontendcss from './tools/builder/frontend-css';
+import admincss from './tools/builder/admin-css';
+import {
+    watch as react_watch,
+    build as react_build
```
Comment (ockham) [inline gulpfile.babel.js]: Nitpick -- let's avoid `snake_case` and use `camelCase` instead (same for other imports/aliases).

File: tools/builder/react.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+		onBuild.bind(
+			this,
+			function( error ) {
+				if ( error ) {
+					log( error );
+					return;
+				}
+				buildStatic( function() {} );
+			}
+		)
```
Comment (ockham) [inline tools/builder/react.js]: Hmm, why the new callback format? The [old one](https://github.com/Automattic/jetpack/pull/9042/files#diff-ba210a9157252cc983268fdc3aa3ed52L218) was a bit easier on the eye :sweat_smile:

File: tools/builder/react.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+			function( error ) {
+				if ( error ) {
+					done( error );
+					return;
+				}
+
+				buildStatic( done );
+			}
+		)
+	);
```
Comment (ockham) [inline tools/builder/react.js]: Same as https://github.com/Automattic/jetpack/pull/9042/files#r225653537

File: tools/builder/react.js (hunk 7 above)
Comment (ockham) [reply tools/builder/react.js]: If it's mainly to invoke `buildStatic` -- should that maybe then become a task of its own?

**Review by ockham — COMMENTED** [review]
Thanks Igor, this is looking pretty good! I left a few more comments. No blockers really except probably for [this](https://github.com/Automattic/jetpack/pull/9042#issuecomment-430332339)

**Comment by zinigor** [comment]
Thanks for catching the languages problem, I have fixed it in the latest commit. Let's not go down the rabbit hole with this one about refactoring React builders? :)

**Comment by ockham** [comment]
> Let's not go down the rabbit hole with this one about refactoring React builders? :)

Yep, okay. I was mainly a bit confused by the [changes](https://github.com/Automattic/jetpack/pull/9042#discussion_r225653537) in callback format, but since it works fine...

**Review by ockham — DISMISSED** [review]
Thanks a bunch @zinigor!

:shipit:

**Comment by kraftbj** [comment]
error `Error: File not found with singular glob: /Users/jetpack/css/jetpack-idc-admin-bar.css (if this was purposeful, use `allowEmpty` option)`

**Comment by ockham** [comment]
Ouch, good catch, @kraftbj! I was able to repro with a fresh clone.

(I'd tried running `yarn distclean` before `yarn build` on my pre-existing clone, but apparently, that didn't remove all built files :thinking: )

**Comment by dereksmart** [comment]
@ockham 
> but apparently, that didn't remove all built files

add `yarn clean-client` to your commands - will do the trick!

**Comment by dereksmart** [comment]
@ockham oh! We should add that directory to the `clean-client` command in a separate PR - I'm happy to do it. 

Side note, I just took a pass at fixing the merge conflict as a result of the changes you made here #10331 - can you take a look and tell me if I got it right?

**Comment by ockham** [comment]
> @ockham oh! We should add that directory to the `clean-client` command in a separate PR - I'm happy to do it. 

Awesome, appreciate it!

> Side note, I just took a pass at fixing the merge conflict as a result of the changes you made here #10331 - can you take a look and tell me if I got it right? 

Thanks! I pushed a commit to remove the `gutenpack` target again which had been accidentally added back.

**Comment by ockham** [comment]
Seems like the original build [error](https://github.com/Automattic/jetpack/pull/9042#issuecomment-430748899) was fixed by the rebase (through https://github.com/Automattic/jetpack/pull/10282/files#diff-1d266473c7938e357a3d2a10bcdf052aL47).

There's still another css file causing the same sort of error:

```
Error: File not found with singular glob: ~/src/jetpack/calypsoify/style-gutenberg.css (if this was purposeful, use `allowEmpty` option)
```

**Comment by dereksmart** [comment]
I think I've fixed it - was able to build clean.  Ready for re-review

`rm -rf css/ && yarn clean-client && yarn distclean && yarn build && yarn watch`

**Comment by kraftbj** [comment]
Builds fine, no errors, but @dereksmart have you checked it out on a live site? The JN site is outputting "undefined" and the app is failing to open.

**Comment by dereksmart** [comment]
@kraftbj I saw that but it was fixed on another rebuild - will look closer tho.

**Comment by dereksmart** [comment]
Alright! made a few small adjustments and seems to be building well now :) 

`yarn clean-client`
`yarn distclean`
`yarn build`
(pause - check to see if everything built properly)
`yarn watch`
(pause - make some changes in the code - should recompile properly)

**Comment by kraftbj** [comment]
Still seeing `undefined` on a JN site.

**Review by kraftbj — APPROVED** [review]
Works when using the right commit :-). Merge if you're happy @dereksmart .

**Pull request merged**